build, ci: Fix linking `bitcoin-chainstate.exe` to `bitcoinkernel.dll` on Windows

hebasto opened the pull request:
On the master branch @ 9a7206a34e3179d7280de98a818a13374178ee7b, linking `bitcoin-chainstate.exe` to `bitcoinkernel.dll` fails:
```
> cmake -B build --preset vs2022-static -DBUILD_UTIL_CHAINSTATE=ON -DBUILD_KERNEL_LIB=ON -DBUILD_SHARED_LIBS=ON
> cmake --build build --config Release -t bitcoin-chainstate
...
LINK : fatal error LNK1181: cannot open input file 'kernel\Release\bitcoinkernel.lib' [C:\Users\hebasto\source\repos\bitcoin\build-static\src\bitcoin-chainstate.vcxproj]
```

This PR fixes this issue and enables the same scenario in the "Win64 native" CI job.

TheCharlatan commented on .github/workflows/ci.yml:
```diff
@@ -182,7 +182,7 @@ jobs:
       - name: Generate build system
         run: |
-          cmake -B build --preset vs2022-static -DCMAKE_TOOLCHAIN_FILE="$env:VCPKG_INSTALLATION_ROOT\scripts\buildsystems\vcpkg.cmake" -DBUILD_GUI=ON -DWITH_BDB=ON -DWITH_MINIUPNPC=ON -DWITH_ZMQ=ON -DBUILD_BENCH=ON -DBUILD_FUZZ_BINARY=ON -DWERROR=ON
+          cmake -B build --preset vs2022-static -DCMAKE_TOOLCHAIN_FILE="$env:VCPKG_INSTALLATION_ROOT\scripts\buildsystems\vcpkg.cmake" -DBUILD_GUI=ON -DWITH_BDB=ON -DWITH_MINIUPNPC=ON -DWITH_ZMQ=ON -DBUILD_BENCH=ON -DBUILD_FUZZ_BINARY=ON -DBUILD_UTIL_CHAINSTATE=ON -DBUILD_KERNEL_LIB=ON -DBUILD_SHARED_LIBS=ON -DWERROR=ON
```
Do you want to take this opportunity to also add a better description to the test name? Maybe something like:
`Win64 native, VS 2022, no depends, libbitcoinkernel`?

hebasto replied:
Thanks! "libbitcoinkernel" taken; it seems awkward to mention "depends", which is a non-Windows thing, on Windows :)

TheCharlatan replied:
Ah, I was thinking of mentioning it because of the macOS build, but I guess we can have native macOS and depends, so it makes sense there, but as you say, does not make sense for msvc.

TheCharlatan commented on src/kernel/cs_main.h:
```diff
@@ -17,6 +17,6 @@
  * chainstate at the same time.
  */
-extern RecursiveMutex cs_main;
+inline RecursiveMutex cs_main;
```
I checked if this could also be a problem for other extern symbols that are currently not used in `bitcoin-chainstate`, but still part of the kernel library. For example in `src/validation.h` we have `extern const std::vector<std::string> CHECKLEVEL_DOC;`, and it looks like using it is indeed problematic too if it were used externally: https://github.com/TheCharlatan/bitcoin/actions/runs/11531552497/job/32102377337#step:10:425.
There are a bunch of other such cases, where I'm not entirely sure how to handle them. Could just ignore them for now, but it is a bit annoying that the library might contain symbols that you couldn't just use.

hebasto commented:
Rebased due to the conflict with the merged https://github.com/bitcoin/bitcoin/pull/31130.

theuni commented:
Concept ACK and code review ACK. This needs a more useful commit message though.

I wasn't familiar with this c++17 functionality, and at first glance it looked scary to me (as though we might get multiple `cs_main`s instantiated).
[
According to cppreference](https://en.cppreference.com/w/cpp/language/inline) though, this is a reasonable (if not more correct/modern) approach.

[This stackoverflow question/answers](https://stackoverflow.com/questions/56876624/inline-stdmutex-in-header-file) also covers our use-case exactly.

Agree with @TheCharlatan that we should track down other cases as well.

theuni commented:
Wait, I take that back. It seems like we could still get 2 separate `cs_main`s here: one in the lib and the other in the binary. ACK retracted while I investigate.

fanquake commented:
Generally I would prefer if we weren't making source code changes to fix (currently edge-case) Windows linking errors; especially if there are more complete fixes we could be making (even if later-on).

It'd be good if all commits explained the problem, and fix. i.e the first commit 4a1accd1df31cba778167b6a3ff9783ffedb46ab `refactor: Inline cs_main definition` has no explanation of what the problem/fix is, and doesn't mention that the change is only needed to fix a build issue when using MSVC, other than "linking fails" in the PR description. I also don't think this should be a labeled "refactor".

hebasto commented:
Rebased. The commit messages have been amended.

@theuni 
> Wait, I take that back. It seems like we could still get 2 separate `cs_main`s here: one in the lib and the other in the binary. ACK retracted while I investigate.

That's an interesting observation. Could you elaborate on how you think the two separate `cs_main` instances might arise?

If there's a relevant section in the documentation or standard that you believe applies, pointing it out would be helpful for clarity.

hebasto commented:
> master currently does not have this issue.

However, the `cs_main` object is duplicated:
```
$ git rev-parse HEAD
94ca99ac51dddbee79d6409ebcc43b1119b0aca9
$ objdump -t build/src/bitcoind | grep cs_main$
0000000000b9d220 g     O .data	0000000000000028              cs_main
$ objdump -t build/src/kernel/libbitcoinkernel.so | grep cs_main$
00000000003dc0a0 g     O .data	0000000000000028              cs_main
```

hebasto commented:
@theuni 
> Wait, I take that back. It seems like we could still get 2 separate `cs_main`s here: one in the lib and the other in the binary. ACK retracted while I investigate.

To clarify this discussion, which binary you are referring to—`bitcoind` or `bitcoin-chainstate`? I assume the latter, as the former uses the static kernel library.

theuni commented:
> @theuni
> 
> > Wait, I take that back. It seems like we could still get 2 separate `cs_main`s here: one in the lib and the other in the binary. ACK retracted while I investigate.
> 
> To clarify this discussion, which binary you are referring to—`bitcoind` or `bitcoin-chainstate`? I assume the latter, as the former uses the static kernel library.

What I'm building isn't really relevant... the combo of the inline var with hidden visibility when built as a shared lib is a potential problem. Even if we have the `CXX_VISIBILITY_PRESET default` hack for now, we'll be removing it in the future.

I'm still trying to understand exactly what the desired outcome is. I suspect that we're going to need default visibility on `cs_main` anyway. So maybe it's not an issue. But I'm still trying to wrap my head around what the `inline` does here when building BOTH the shared lib and the downstream app that way.

I also want to understand if this is just a distraction, and if we can just keep things the way they currently are, but fixup the visibilities.

hebasto commented:
> What I'm building isn't really relevant... the combo of the inline var with hidden visibility when built as a shared lib is a potential problem.

I think it could be problematic when an inline variable with hidden visibility is part of a static library, which is then linked both to a binary and to a shared library. If the binary does not share dependencies with the shared library, it should be safe.

theuni commented:
From the docs

> The module definition file will be passed to the linker causing all symbols to be exported from the .dll. **For global data symbols, __declspec(dllimport) must still be used when compiling against the code in the .dll.** All other function symbols will be automatically exported and imported by callers

See bold text. I suspect that's the real issue here, and that inlining is the opposite of what we want.

theuni commented:
> If the binary does not share dependencies with the shared library, it should be safe.

The binary has to find `cs_main` somewhere though. And we don't want it adding its own.

I think I understand what's happening now and I'm testing an alternative.

hebasto commented:
> > If the binary does not share dependencies with the shared library, it should be safe.
> 
> The binary has to find `cs_main` somewhere though.

Right. `bitcoin-chainstate` finds it in `libbitcoinkernel.so`.

hebasto commented:
> From the docs
> 
> > The module definition file will be passed to the linker causing all symbols to be exported from the .dll. **For global data symbols, __declspec(dllimport) must still be used when compiling against the code in the .dll.** All other function symbols will be automatically exported and imported by callers
> 
> See bold text. I suspect that's the real issue here, and that inlining is the opposite of what we want.

Mind sharing a link?

theuni commented:
Please give https://github.com/theuni/bitcoin/commits/pr31158/ a shot. It's not meant to be robust, but it should illustrate the fix (using dllimport for bitcoin-chainstate.exe).

hebasto commented:
> Please give https://github.com/theuni/bitcoin/commits/pr31158/ a shot. It's not meant to be robust, but it should illustrate the fix (using dllimport for bitcoin-chainstate.exe).

CI fails: https://github.com/theuni/bitcoin/actions/runs/13163570761/job/36738155052

theuni commented:
> > Please give https://github.com/theuni/bitcoin/commits/pr31158/ a shot. It's not meant to be robust, but it should illustrate the fix (using dllimport for bitcoin-chainstate.exe).
> 
> CI fails: https://github.com/theuni/bitcoin/actions/runs/13163570761/job/36738155052

Yeah, the above is only enough to test `bitcoin-chainstate.exe`. In reality, the condition in the header would need to be something like:
`#elif defined(MSC_VER) && !defined(STATIC_LIBBITCOINKERNEL) && !defined(BITCOIN_INTERNAL_BUILD)`, which would get defined for everything that's not bitcoin-chainstate.

hebasto commented:
@theuni 

> Please give https://github.com/theuni/bitcoin/commits/pr31158/ a shot. It's not meant to be robust, but it should illustrate the fix (using dllimport for bitcoin-chainstate.exe).

Thank you for your insight!

This PR has been reworked based on your branch. The `inline` variables are avoided.

theuni commented:
Thank you @hebasto! We had a pretty long discussion about this. I think that hebasto's approach is probably fine and would never be an issue in the real world. I'll admit that I'm being overly paranoid and probably wrong about the guarantees that linkers provide...

But still, I'm not alone in my paranoia. For example, Google has [banned non-constexpr inline header variables](https://groups.google.com/a/chromium.org/g/cxx/c/hmyGFD80ocE/m/qVe-hqAVDAAJ) for Chromium out of similar concerns.

[This reddit thread](https://www.reddit.com/r/cpp/comments/ulhv0b/low_performance_of_inline_global_variables/) also contains a discussion from which illustrates some of my concerns and made me nervous.

For reference, he's how the binaries changed.

extern (as-in current master):
```
 $ nm -C src/kernel/libbitcoinkernel.so | grep cs_main
00000000003b3028 D cs_main

 $ nm -C  src/bitcoin-chainstate | grep cs_main
                 U cs_main
```


header-inline (this PR's original proposal):
```
 $ nm -C src/kernel/libbitcoinkernel.so | grep cs_main
00000000003b3010 V cs_main
00000000003b35f0 V guard variable for cs_main

 $ nm -C  src/bitcoin-chainstate | grep cs_main
000000000000e010 V cs_main
000000000000e048 V guard variable for cs_main
```
---
The root cause here is that CMake's visibility machinery handles functions but not variables. So there are basically 3 potential solutions:

- A c++17-style inline variable as hebasto originally proposed here. I admit I'm being overly paranoid, but I'd rather not do that.
- Give cs_main hidden visiblility and add a `get_csmain()` getter function. That would work but would mean that bitcoin-chainstate's code would diverge from the rest of the codebase. (assuming we didn't do a global replacement, which would be a HUGE hammer)
- Add our own visibility header for exported variables, as hebasto has done here in the updated impl of this PR.

I pretty strongly prefer the 3rd as it's minimally invasive and leaves no potential for regressions in bitcoind/kernel.

Will review in detail.

theuni commented on src/kernel/symbol_visibility.h:
```diff
@@ -0,0 +1,18 @@
+  #else
+    #define BITCOINKERNEL_EXPORT __attribute__ ((visibility ("default")))
+  #endif
+#elif defined(_WIN32) && !defined(BITCOINKERNEL_STATIC)
```
The meaning of `BITCOINKERNEL_STATIC` here is overloaded. Just to differentiate the use-cases, I'd prefer to have a 3rd define here:

`#elif defined(_WIN32) && !defined(BITCOIN_BUILD_INTERNAL) && !defined(BITCOINKERNEL_STATIC)`

`BITCOINKERNEL_STATIC`: I'm a user of the kernel building against a static kernel.
`BITCOIN_BUILD_INTERNAL`: I'm a component of Core who's not building against the kernel at all, I just happen to be seeing this header.

The former should stay as-is for bitcoin-chainstate, the latter is what we should be defining for the other internal libs.

hebasto replied:
> `BITCOIN_BUILD_INTERNAL`: I'm a component of Core who's not building against the kernel at all, I just happen to be seeing this header.

`BITCOIN_INTERNAL_BUILD`  was in my draft of the updated branch this morning :)

Will rework the PR shortly.

hebasto replied:
Thanks! Reworked.

hebasto committed:
cmake, kernel: Enable `WINDOWS_EXPORT_ALL_SYMBOLS` property

To enable linking to `bitcoinkernel.dll`, temporarily enable the CMake's
`WINDOWS_EXPORT_ALL_SYMBOLS` property. Global data symbols, such as
`cs_main`, must still be properly attributed with
`__declspec(dllexport)` and `__declspec(dllimport)`.

This change can be reconsidered once the kernel API headers become
available.

Co-authored-by: Cory Fields <cory-nospam-@coryfields.com>

hebasto committed:
ci, windows: Build shared and static `bitcoinkernel` libraries

This change expands the matrix jobs for the native Windows builds to
test shared and static `bitcoinkernel` libraries.
To conserve resources, the functional tests are enabled for a single job
only.

hebasto committed:
ci, windows: Treat linker warnings as errors

This change prevents issues related to the `__declspec(dllimport)`
attribute, such as LNK4217 or LNK4286.

hodlinator commented:
Ran into this linker error when attempting to test #31161 on Windows yesterday.
Thanks for working on this, good to see linker warnings as errors enabled in CI!

The kernel library exposing `cs_main` instead of exposing functions that automatically take the correct locks makes me want to cry. But we have to deal with where we are.

Concept ACK.

hodlinator commented on src/kernel/CMakeLists.txt:
```diff
@@ -90,14 +90,22 @@ target_link_libraries(bitcoinkernel
+  # Note this is a quick hack that will be removed as we
+  # incrementally define what to export from the library.
   CXX_VISIBILITY_PRESET default
+  # Temporarily enable WINDOWS_EXPORT_ALL_SYMBOLS until a dedicated
+  # header with __declspec(dllexport) directives is available.
+  WINDOWS_EXPORT_ALL_SYMBOLS TRUE
```
Isn't *symbol_visibility.h* this dedicated header?

TheCharlatan commented:
I'm not sure about the current approach of this change. I think we could go the way of manually exporting symbols in our c++ code, but then I would call into question the need for this project to ship a C header in the first place. Having two symbol-exporting headers, one for our internal code, and one for a potential future C header seems very confusing. 

Also, as maflcko already said, we could just not directly use these globals from code that links to the shared library. If I understand the problem correctly, in the case of cs_main the required change is very small, and taking a similar approach for the other ones (if we ever run into them) seems reasonable.

fanquake commented:
> I'm not sure about the current approach of this change.

I agree. There's no rush or need to do anything here, and doing something just to "make things green/compile" seems like the wrong choice, rather than making some good design decisions.

hebasto commented:
Converted to a draft for now.
